### PR TITLE
rtl_tcp server UI: clients list, listener cap, auth toggle (#395)

### LINKED ISSUE #395: rtl_tcp: server UI — client list + cap + auth toggle
Part of #390 (sub-issue 5). Blocked on #391, #392, #394.

## Summary

Add server-side UI for the multi-client features: live client list with role badges, listener cap setting, auth toggle + key display. Lives in the existing rtl_tcp server panel in the Linux sidebar (macOS SwiftUI port is a separate follow-up).

## Widgets

Under the existing rtl_tcp server panel (tracked by #303):

### "Connected clients" subsection

Live-updating list. One row per connected client:
- Peer address (host:port)
- Role badge (Controller / Listener) — accent color for controller, neutral for listener
- Connection duration (human-readable: "2m 14s")
- Drop counter (for listeners — shows how many chunks were dropped due to slow reads, zero is the usual value)
- For the controller row: no special controls. Removing a controller mid-session via UI is sharper-edged than we want; users take control from the other end via #393.

When no clients connected: empty-state placeholder "No clients connected".

### "Listener cap" spin row

Default 10. Range [0, 32]. 0 = "control only — no listeners allowed". Saved to `sdr_config` per-server-instance. Change takes effect on next accept; existing listeners are not kicked when the cap is lowered (deliberate — surprise disconnection is rude).

### "Require key" switch row + key display

Switch: off by default (matches current behavior). When switched on:
- Server generates + persists a key (#394).
- Key display row appears below with:
  - Masked key by default (`••••••••…•••`)
  - "Reveal" toggle button → shows the full key
  - "Copy" button → copies the key to clipboard
  - "Regenerate" button → invalidates the current key, generates a new one

When switched off: key row disappears (key retained in keyring so flipping back doesn't regenerate).

## Events driving the live list

New `DspToUi` variant: `ServerClientListChanged(Vec<ServerClient>)` emitted from the server on every accept / disconnect. UI handler replaces the listbox contents.

Actually, a poll-based pattern might be simpler: the server exposes `stats.clients()` returning the current list; UI has a 1-second `glib::timeout_add_local` that reads stats and updates the listbox only if the set changed. Chooses polling over event-driven to avoid a hot-path dispatch every time a listener reconnects.

Either pattern is fine; pick whichever fits the existing stats pipeline better.

## Where to look

- `crates/sdr-ui/src/sidebar/server_panel.rs` — the existing server panel. New widgets go here.
- `crates/sdr-server-rtltcp/src/server.rs` — `ServerStats` extension to include per-client info.
- `crates/sdr-core/src/messages.rs` — potentially new `DspToUi` variant for client-list updates (if event-driven).
- `crates/sdr-ui/src/window.rs` — `handle_dsp_message` wiring.

## Acceptance

- [ ] Connected-clients list renders live. Each row shows peer address, role, duration, drops.
- [ ] Listener cap spin row functional, persisted via `sdr_config`, applied on next accept.
- [ ] Auth toggle shows/hides the key row; persists.
- [ ] Copy button sends the key to the clipboard.
- [ ] Reveal toggle shows/masks the key in the UI.
- [ ] Regenerate button produces a new key + invalidates the old one (next reconnect with old key fails).
- [ ] Smoke test: two clients connect (one control, one listener), server panel shows both with correct badges, cap-row change is persisted across restart.

## Blocked on

- #391 — fan-out (so listener concept exists).
- #392 — roles (so the badges are meaningful).
- #394 — auth (so the toggle does anything).

## Related

- #390 — parent epic.
- #303 — existing rtl_tcp server panel (this builds on top of it).

### PR DESCRIPTION
**Pull request by jasonherald**
## Summary

Closes #395 — fifth sub-issue of multi-client epic #390. Adds the server-side UI for the features shipped across #391-#394: live listener cap knob, pre-shared-key auth toggle with keyring storage, and a connected-clients list with per-client role badges, duration, and drop counters. Five implementation commits, no wire/protocol changes (all backend surface was already there from PRs #402-#405).

Behind the scenes, the server crate grows a minimal live-update API (`Server::set_listener_cap` / `set_auth_key`) so UI edits take effect on the next accept without restarting (and kicking all connected clients). The mDNS `TxtRecord` gains an `auth_required` field so discovery clients can prompt for a key BEFORE dispatching connect.

## Commits

1. **`a6f41c7`** — `Server::set_listener_cap(n)` + `set_auth_key(key)` backed by `Arc<AtomicUsize>` / `Arc<Mutex<Option<Vec<u8>>>>`. Shared `validate_auth_key_length` helper so the live-update surface cannot accept inputs `Server::start` would refuse. Accept + handshake paths snapshot their respective values once per connection. 4 new unit tests for the validator.
2. **`837853c`** — `TxtRecord.auth_required: Option<bool>`. `Some(true)` emits `auth_required=true`; everything else omits. Parser accepts `true` / `1` / `yes` case-insensitively. CLI and GTK advertisers wired via new `Server::auth_required()` accessor. 4 new tests.
3. **`690af0e`** — Listener cap `AdwSpinRow`, range [0, 32], default 10. Config persistence + live-apply on change. Intentionally NOT in `set_controls_locked` so the user can tune the cap without restarting.
4. **`c707ca0`** — "Require key" `AdwSwitchRow` + `auth_key_row` with masked/reveal subtitle, Copy, Regenerate. Keyring storage via `KEYRING_KEY_AUTH_KEY` as lowercase hex. `ensure_server_auth_key()` load-or-generate. 3 new tests for the hex helpers (round-trip, malformed rejection incl. non-ASCII emoji, empty-input edge case).
5. **`bfaa415`** — "Connected clients" expander with per-client `AdwActionRow`: peer title, `{role} · {duration} · {drops}` subtitle, colored dot prefix (accent for Controller, dim for Listener). Diff by sorted `ClientId` hash so the ListBox rebuilds only on accept/disconnect (user scroll / hover state survive poll cadence).

## Live-update design

Listener cap and auth key are both specified as live-updatable on a running server (issue says "change takes effect on next accept"). New `Server` methods:
- `set_listener_cap(n: usize)` — `Relaxed` store, accept thread reads per admission
- `set_auth_key(key: Option<Vec<u8>>) -> Result<(), ServerError>` — validates length, then mutex-stored, handshake snapshots once per connection
- `set_auth_key` uses the same `validate_auth_key_length` helper as `Server::start` so construction and live-update enforce identical `None | 1..=256` rule.

`spawn_client_workers` snapshots `auth_key` **once** at entry so a mid-handshake `set_auth_key` can't split an eager client's key bytes (written against the old value) from the server's validation (running against the new value). Each handshake is bound to a single key view.

## Test plan

- [ ] `make install CARGO_FLAGS="--release --features whisper-cuda"` — build clean (verified locally)
- [ ] Panel visibility: starts hidden, appears when non-RTLSDR source is selected with a dongle plugged in
- [ ] Listener cap: spin up/down; setting 0 + connecting a listener → server denies with `ListenerCapReached`
- [ ] Listener cap live: with a listener already connected, drop the cap to 0 → listener stays, next listener gets denied
- [ ] Auth toggle on: key row appears with masked subtitle
- [ ] Auth reveal button: flips subtitle between masked and 64-char hex; icon flips
- [ ] Auth copy: clipboard contains the full hex; toast confirms
- [ ] Auth regenerate: subtitle updates; old key fails on the next reconnect (test with `sdr-rtl-tcp --auth-key <old-hex>`); toast confirms
- [ ] Auth toggle off: key row hides; keyring entry preserved (toggle back on reloads same key)
- [ ] Connected clients list: connect one controller + one listener with a vanilla client (e.g. `nc localhost 1234`) and our client → both rows render with correct role badge and duration
- [ ] Clients list empty state: "No clients connected" row when no one connected
- [ ] Server restart across session: toggle auth off → stop → start → toggle auth back on → same keyring key surfaces
- [ ] mDNS TXT: with auth on, another instance's discovery picker should show "auth_required" in the entry (future #396 UI will surface this; today, grep `avahi-browse -r _rtl_tcp._tcp` for `auth_required=true`)

## Remaining epic #390 sub-issues

- **#396** — client UI (role picker + takeover button + kicked toast + auth key field) — next PR
- **#397** — FIPS-compliant transport encryption (investigation)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Server auth key management: reveal, copy, regenerate with OS keyring-backed persistence and graceful fallback.
  * Live-adjustable listener capacity with immediate application.
  * Connected clients list in server panel with per-client details.
  * Discovery now advertises whether authentication is required and updates live.

* **Quality / Tests**
  * Unit and integration tests updated to validate TXT record handling, auth-key import/export rules, and live-update behavior.

* **Validation**
  * Strict lowercase-hex validation for auth-key import/export.

### CODE REVIEW

File: crates/sdr-server-rtltcp/src/server.rs [hunk 1]
```diff
@@ -456,10 +471,74 @@ impl Server {
...
+    /// iff [`Server::set_auth_key`] has been called with `Some(_)`
+    /// (or the starting `ServerConfig.auth_key` was `Some`). Does
+    /// not leak the key itself — useful for stamping the mDNS TXT
+    /// `auth_required=true` field without handing the caller the
+    /// raw key bytes. Returns `false` on a poisoned mutex because
+    /// an unknown auth state should default to the safer "advertise
+    /// no auth" so clients don't falsely prompt for a key. Per #395.
+    pub fn auth_required(&self) -> bool {
+        self.auth_key.lock().is_ok_and(|g| g.is_some())
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-server-rtltcp/src/server.rs:580]: Fixed in a62ebdc. Added `auth_required_cache: Arc<AtomicBool>` to `Server` — seeded from `ServerConfig.auth_key.is_some()` at construction, updated inside `set_auth_key` on every successful mutation. `auth_required()` now falls back to the cache on `PoisonError` with a `warn!` log so the mDNS TXT stays honest when the mutex is unusable. The handshake path and `set_auth_key` keep their fatal-on-poison treatment unchanged.

File: crates/sdr-ui/src/sidebar/server_panel.rs [hunk 2]
```diff
@@ -64,6 +86,68 @@ const SERVER_PORT_STEP: f64 = 1.0;
...
+        let hi = char::from(chunk[0]).to_digit(16)?;
+        let lo = char::from(chunk[1]).to_digit(16)?;
+        // `hi` and `lo` are each 0..=15 (validated by `to_digit(16)`),
+        // so `(hi << 4) | lo` fits in u8 with the top 24 bits zero —
+        // `u8::try_from` is infallible here but keeps clippy's
+        // `cast_possible_truncation` quiet.
+        let byte = u8::try_from((hi << 4) | lo).ok()?;
+        out.push(byte);
+    }
+    Some(out)
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/sidebar/server_panel.rs:162]: Fixed in a62ebdc. `auth_key_from_hex` now enforces both ends of the range: rejects empty strings (unchanged) AND any hex input longer than `2 * MAX_AUTH_KEY_LEN` chars so a corrupt / oversize keyring entry returns `None` rather than decoding to an oversize `Vec<u8>` that fails `Server::start`'s length gate later. Added regression test `auth_key_from_hex_rejects_oversize_decoded_length` that pins exactly-at-cap as decodes, cap+1 as rejects.

File: crates/sdr-ui/src/sidebar/server_panel.rs [hunk 3]
```diff
@@ -535,6 +707,85 @@ pub fn build_server_panel() -> ServerPanel {
...
+        .tooltip_text("Copy key to clipboard")
+        .valign(gtk4::Align::Center)
+        .css_classes(["flat"])
+        .build();
+    let auth_key_regenerate_button = gtk4::Button::builder()
+        .icon_name("view-refresh-symbolic")
+        .tooltip_text("Regenerate key — old key stops working for future reconnects")
+        .valign(gtk4::Align::Center)
+        .css_classes(["flat"])
+        .build();
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/sidebar/server_panel.rs:807]: Fixed in a62ebdc. Added `accessible::Property::Label` on all three buttons in `server_panel.rs` — "Reveal key", "Copy key to clipboard", "Regenerate key". In `window.rs` the reveal button's accessible label is flipped alongside `icon_name` + `tooltip_text` in BOTH directions: reveal-click ↔ hide, and the toggle-on reset path that forces everything back to the masked state. Screen readers now announce the current action, not the stale build-time label.

File: crates/sdr-ui/src/window.rs [hunk 4]
```diff
@@ -3082,6 +3168,51 @@ fn connect_share_switch(
...
+    let auth_key_revealed: Rc<std::cell::Cell<bool>> = Rc::new(std::cell::Cell::new(false));
+
+    // If auth was restored as ON from config, eagerly load the
+    // key from the keyring so the key row reflects real state
+    // before the user interacts with anything. The server isn't
+    // running yet (that requires the share_row flip), so no
+    // `set_auth_key` call here — just UI state.
+    if panels.server.auth_require_row.is_active() {
+        let key = ensure_server_auth_key();
+        *current_auth_key.borrow_mut() = Some(key);
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs:3211]: Fixed in a62ebdc. `build_server_config_from_panel` now takes `pending_auth_key: Option<Vec<u8>>` from the caller instead of calling `ensure_server_auth_key()` internally. The share_row active-notify handler reads from `current_auth_key` — the same cell the Reveal / Copy buttons dereference — so there's exactly one `ensure_server_auth_key` call per enable-auth flow, and the Reveal-button-displayed value is guaranteed to match what `Server::start` receives (even in the keyring-IO-unavailable case that triggered this finding).

File: crates/sdr-ui/src/window.rs [hunk 5]
```diff
@@ -3187,11 +3319,210 @@ fn connect_share_switch(
...
+                {
+                    tracing::warn!(%e, "Server::set_auth_key(None) failed on toggle-off");
+                }
+                *current_key_for_toggle.borrow_mut() = None;
+                key_row.set_visible(false);
+                // Zero the revealed flag too so a next toggle-on
+                // starts masked regardless of the prior reveal
+                // state.
+                revealed_for_toggle.set(false);
+            }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs:3465]: Fixed in a62ebdc. Refactored the toggle handler around a new `apply_live_auth_change()` helper:
1. Lock the `RunningServer` cell.
2. Call `Server::set_auth_key(new)` — on `Err`, toast + return `false`.
3. On `Ok`, refresh the mDNS advertiser (fix 6) and return `true`.

Caller gates UI mutation on that `bool`: UI state only advances when the server actually holds the new value. If the helper returns `false`, the handler reverts the switch via a new `auth_toggle_reentry_guard: Rc<Cell<bool>>` (same pattern as the share_row guard) so `row.set_active(prev)` doesn't re-fire the handler.

Same treatment applied to the regenerate-button click handler: keyring-write failure stays tolerable (in-memory fallback), but a `set_auth_key` failure now toasts + returns without advancing `current_auth_key`, the displayed subtitle, or the "New key generated" confirmation toast.

**Comment by jasonherald** [comment]
**Outside-diff fix (advertiser refresh on auth toggle):**

Fixed in a62ebdc. New `refresh_advertiser_for_auth_change()` helper tears down the existing `Advertiser` (letting its `Drop` unregister first) and re-announces via `build_advertiser()`, which reads the current `Server::auth_required()` state. Called from the new `apply_live_auth_change()` helper after every successful `set_auth_key`. Honors the user's `advertise_row` preference — we don't bring advertising back online just because auth flipped.

Regenerate intentionally does NOT refresh the advertiser because `auth_required` stays `true` before and after a key rotation, so the TXT is unchanged and re-announcing under the same instance name is pure churn.

On rebuild failure: toast + leave `advertiser = None`, same pattern as the initial-start failure path. Server keeps running without a fresh TXT.

File: crates/sdr-ui/src/window.rs [hunk 6]
```diff
@@ -3303,6 +3744,11 @@ fn connect_server_status_polling(
     // is cheap but clearing the ListBox resets any user scroll
     // position, so we short-circuit on unchanged ticks.
     let last_activity_key: Rc<Cell<(usize, Option<Instant>)>> = Rc::new(Cell::new((0, None)));
+    // Clients-list diff key: hash of the connected `ClientId` set.
+    // Same rationale as `last_activity_key` — rebuilding the
+    // ListBox every tick would wipe user scroll position + flash
+    // the hover highlight off each 500 ms tick. Per issue #395.
+    let last_clients_key: Rc<Cell<u64>> = Rc::new(Cell::new(0));
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs]: Fixed in 6b8804b. Two parts:

1. **Wider hash** — `render_clients_list`'s diff key now hashes `(id, peer, role_discriminant, buffers_dropped, elapsed_secs)` per client (sorted by id first for stable ordering). Ticking uptime or incrementing drops bump the hash, so a stable connected set still re-renders once per second. Rebuild cost at 2 Hz × N ≤ 32 is trivial vs. the USB/DSP hot path.

2. **Empty-set cache invalidation** — `last_clients_key` is now `Rc<Cell<Option<u64>>>` with `None` as the initial / invalidated sentinel. Instead of plumbing a cache-invalidate call across the `connect_share_switch` / `connect_server_status_polling` function boundary (where `reset_clients_list` lives in the former and the cache in the latter), `render_clients_list` itself detects the reset state via `widgets.clients_list.first_child().is_none()` — the empty-state render always appends a placeholder row, so zero children means the reset path cleared it between ticks. Forces a rebuild even on a matching hash; subsequent renders see the placeholder row and proceed normally.

Also: the `now` `Instant` is captured once and reused for both the hashed `elapsed_secs` and the displayed `format_uptime()` string, so the hash can't match while the rendered duration claims a different second.

File: crates/sdr-ui/src/window.rs [hunk 7]
```diff
@@ -3187,11 +3347,284 @@ fn connect_share_switch(
...
+                && let Err(e) = handle.server.set_auth_key(Some(fresh.clone()))
+            {
+                tracing::warn!(%e, "Server::set_auth_key failed on regenerate");
+                if let Some(overlay) = toast_overlay_for_regen.upgrade() {
+                    overlay.add_toast(adw::Toast::new(&format!(
+                        "Couldn't apply new key to the running server: {e}"
+                    )));
+                }
+                return;
+            }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs]: Fixed in 05f9ac6. Reordered the regenerate handler to the same contract the toggle path uses: server-apply first, persist second, UI third.

1. `apply_live_auth_change(&running_for_auth_regen, Some(fresh.clone()), None, &toast_overlay_for_regen)` — if it returns `false` (server borrow busy OR `set_auth_key` errored), toast has already fired inside the helper; we just `return` without touching keyring or UI.
2. `save_server_auth_key_to_keyring(&fresh)` — failure here stays tolerable (current session already works via the in-memory key that the server accepted; next launch falls back to the old key until the user re-clicks Regenerate).
3. UI mutation (`current_auth_key`, subtitle, success toast).

Passing `widgets: None` to `apply_live_auth_change` skips the advertiser rebuild — regenerate keeps `auth_required = true` so the TXT doesn't change, and re-announcing under the same instance name is pure churn.

File: crates/sdr-ui/src/window.rs [hunk 8]
```diff
@@ -3661,6 +4116,149 @@ fn render_activity_log(
...
+/// state: single "No clients connected" placeholder row plus
+/// matching expander subtitle. Rebuilds only when the
+/// `ClientId` set changes (i.e. a client joined or left) so
+/// user scroll position + hover highlight survive the 500 ms
+/// poll cadence. The per-client field churn (duration bumps,
+/// drop-count increments) IS NOT diffed here — we'd have to
+/// index rows by id to patch them in place, which is more
+/// complexity than the row-per-client design needs. Callers
+/// who want a running clock on each row should watch
+/// individual slots via a dedicated widget. Per issue #395.
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs]: Fixed in 05f9ac6. Updated the docstring to describe both the current behavior:
- **Rebuild trigger**: hashes `(id, peer, role, drops, elapsed_secs)` so uptime/drops bumps trigger rebuilds, not just accept/disconnect.
- **Stop/start invalidation**: `first_child().is_none()` detection so the empty→empty transition across stop/start still repaints the placeholder.

Good catch — a stale contract in the doc would have invited exactly the kind of "the diff key only watches id-set" regression this round just fixed.
